### ISSUE TITLE
False positive warning `warning: instantiation of variable 'declvar<A>' required here, but no definition is available`, when the definition is not required

### ISSUE THREAD

**Issue by HolyBlackCat** [llvm/llvm-project#128657]
Consider the following code:
```cpp
#include <utility>

struct A {explicit constexpr operator bool() const {return true;}};
struct B {int x = 1; explicit constexpr operator bool() const {return x;}};

template <typename T>
extern T declvar;

template <typename T>
concept C = requires{typename std::bool_constant<declvar<T> ? true : false>;};

static_assert(C<A>);
static_assert(!C<B>);
```
This works correctly on all 3 bug compilers including Clang (and seems to be the only good way of doing this).

But Clang emits a useless warning:
```none
<source>:10:50: warning: instantiation of variable 'declvar<A>' required here, but no definition is available [-Wundefined-var-template]
   10 | concept C = requires{typename std::bool_constant<declvar<T> ? true : false>;};
      |                                                  ^
<source>:7:10: note: forward declaration of template entity is here
    7 | extern T declvar;
      |          ^
<source>:10:50: note: add an explicit instantiation declaration to suppress this warning if 'declvar<A>' is explicitly instantiated in another translation unit
   10 | concept C = requires{typename std::bool_constant<declvar<T> ? true : false>;};
      |                                                  ^
```
The variable definition isn't needed here, so it shouldn't warn.

**Comment by frederick-vs-ja** [comment]
I _think_ `declvar<A>` & `declvar<B>` are still odr-used per [[basic.def.odr]/7](https://eel.is/c++draft/basic.def.odr#7), even though the odr-uses are constant-evaluation-only.

**Comment by HolyBlackCat** [comment]
> I _think_ `declvar<A>` & `declvar<B>` are still odr-used per [[basic.def.odr]/7](https://eel.is/c++draft/basic.def.odr#7), even though the odr-uses are constant-evaluation-only.

Even if they technically are, I think Clang shouldn't warn if there's no linker error.

**Comment by frederick-vs-ja** [comment]
> > I _think_ `declvar<A>` & `declvar<B>` are still odr-used per [[basic.def.odr]/7](https://eel.is/c++draft/basic.def.odr#7), even though the odr-uses are constant-evaluation-only.
> 
> Even if they technically are, I think Clang shouldn't warn if there's no linker error.

Hmm, it seems to me that we should figure out "constant-evaluation-only odr-use", e.g.:

```C++
extern int x;
constexpr bool to_truth(auto&&) { return true; }
static_assert(::to_truth(x)); // Odr-use!
// But it can be statically known that such odr-use won't escape from constant evaluation.
```

It seems OK to exclude such use from odr-use...

**Comment by frederick-vs-ja** [comment]
> it seems to me that we should figure out "constant-evaluation-only odr-use"

I guess such a thing is already necessary for C++. Consider function declaration `auto fun(Arg arg) -> Ret;`, we can odr-use `arg` in `Ret` without actually creating parameter object/reference, as long as any such odr-use is resolved in constant evaluation and doesn't escape.
